src/widgets/table.rs

```diff
@@ -751,25 +751,25 @@ mod tests {
             // without selection, more than needed width
             assert_eq!(
                 get_test(&[Constraint::Ratio(1, 3), Constraint::Ratio(1, 3)], 20, 0),
-                &[(0, 6), (7, 6)]
+                &[(0, 7), (8, 6)]
             );
 
             // with selection, more than needed width
             assert_eq!(
                 get_test(&[Constraint::Ratio(1, 3), Constraint::Ratio(1, 3)], 20, 3),
-                &[(3, 6), (10, 6)]
+                &[(3, 7), (11, 6)]
             );
```

Comment:
something is not right here, they should be the same width (there is more than available space)

Reply:
I think this test is not deterministic? I stashed all my changes and ran this test on `main`; and I'm always getting a failure.

<img width="755" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/b8b012f4-7204-45a5-ad63-b00b29556c1c">

Reply:
When I run on `main` with these two changes:

```rust
            dbg!(changes.get(&element.start));
            dbg!(changes.get(&element.end));
```

```
$ cargo test -- widgets::table::tests::table_column_widths::ratio_constraint --nocapture
running 1 test
[src/layout.rs:568] changes.get(&element.start) = None
[src/layout.rs:569] changes.get(&element.end) = Some(
    0.0,
)
[src/layout.rs:568] changes.get(&element.start) = Some(
    0.0,
)
[src/layout.rs:569] changes.get(&element.end) = Some(
    6.666666666666666,
)
[src/layout.rs:568] changes.get(&element.start) = Some(
    6.666666666666666,
)
[src/layout.rs:569] changes.get(&element.end) = Some(
    7.666666666666666,
)
[src/layout.rs:568] changes.get(&element.start) = Some(
    7.666666666666666,
)
[src/layout.rs:569] changes.get(&element.end) = Some(
    14.333333333333332,
)
thread 'widgets::table::tests::table_column_widths::ratio_constraint' panicked at 'assertion failed: `(left == right)`
  left: `[(0, 7), (8, 6)]`,
 right: `[(0, 6), (7, 6)]`', src/widgets/table.rs:752:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test widgets::table::tests::table_column_widths::ratio_constraint ... FAILED

failures:

failures:
    widgets::table::tests::table_column_widths::ratio_constraint
```

I think in #404 there was no rounding, so your tests were passing then. But #408 introduced rounding.

There's something I don't understand though. Here's the git log history:

<img width="520" alt="image" src="https://github.com/ratatui-org/ratatui/assets/1813121/f2c38cc9-2bea-47ae-82a1-37d7962269e7">

You can see that #408 was before #404. After merging #404 to `main`, the tests were indeed failing: https://github.com/ratatui-org/ratatui/actions/runs/5911487064/job/16034036047.

But my understanding was that GitHub's merge queue would run the tests again on the latest HEAD before merging; which would have caught this issue before merging #404. Does anyone know why that did not happen? Is there a feature that we need to enable on the merge queue feature on GitHub?

Reply:
> You can see that https://github.com/ratatui-org/ratatui/pull/408 was before https://github.com/ratatui-org/ratatui/pull/404. And after merging https://github.com/ratatui-org/ratatui/pull/404 to main, the tests were indeed failing on CI: https://github.com/ratatui-org/ratatui/actions/runs/5911487064/job/16034036047.

after some quick tests, i came to the same conclusion

> But my understanding was that GitHub's merge queue would run the tests again on the latest HEAD before merging; which would have caught this issue before merging https://github.com/ratatui-org/ratatui/pull/404. Does anyone know why that did not happen? Is there a feature that we need to enable on the merge queue feature on GitHub?

i was also under that assumption, but maybe the branch-protection on `main` does not require all tests to pass? i came to this conclusion because of the following doc snippit

> Once a pull request has passed all required branch protection checks, a user with write access to the repository can add the pull request to the queue. The merge queue will ensure the pull request's changes pass all required status checks when applied to the latest version of the target branch and any pull requests already in the queue.

another possibility is that maybe because [of these lines](https://github.com/ratatui-org/ratatui/blob/b6b2da5eb761ac5894cc7a2ee67f422312b63cfc/.github/workflows/ci.yml#L34-L38) the *pull request head* was checked out *instead of the merged head*, but i am not sure, because i dont know if [this run](https://github.com/ratatui-org/ratatui/actions/runs/5911487064) is the run that tested the merge-ability, or the test after the merge

Reply:
for now i have created a meta-issue to track this instead of this PR https://github.com/ratatui-org/ratatui/issues/418

Reply:
Great, I'll resolve this conversation.

Reply:
I fixed the table tests that were broken in https://github.com/ratatui-org/ratatui/pull/419